Split wallbox sensor type and number type

hesselonline opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Seperate NUMBER_TYPES dict in const.py

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

hesselonline committed:
Split sensor type and number type in const

hesselonline committed:
review comments

hesselonline committed:
moved sensor_types and number_types to components

hesselonline committed:
fix pylint error

hesselonline commented:
@MartinHjelmare @cdce8p this is what I meant. Creating a seperate NUMBER_TYPES dict in const which determines the number entities to be added. This also allows other number components to be added in the future (or the related sensor component to be removed in the future). Furthermore, the code is in many ways similar between number.py and sensor.py which I find makes it easy to read.

MartinHjelmare commented on homeassistant/components/wallbox/const.py:
```diff
@@ -43,6 +44,23 @@ class WallboxSensorEntityDescription(SensorEntityDescription):
+    min_value: float = 0
+
+
+NUMBER_TYPES: dict[str, NumberEntityDescription] = {
```
Please move this to the number platform.

MartinHjelmare commented on homeassistant/components/wallbox/number.py:
```diff
@@ -23,20 +28,28 @@ async def async_setup_entry(hass, config, async_add_entities):
+        async_add_entities(
+            [
+                WallboxNumber(coordinator, config, description)
+                for ent in filtered_data
```
We don't need to pre-filter the coordinator data. It's enough to check if there's a description for the data key. Use `dict.get` to look up the key in the number types dict.

hesselonline replied:
OK, wil change this in new commit

MartinHjelmare commented on homeassistant/components/wallbox/number.py:
```diff
@@ -23,20 +28,28 @@ async def async_setup_entry(hass, config, async_add_entities):
-        self._attr_device_class = sensor_description.device_class
+        self._attr_name = f"{config.title} {description.name}"
+        self._attr_min_value = description.min_value
+        self._attr_device_class = description.device_class
```
We don't need this since it's set by the description.
```suggestion
```

MartinHjelmare commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -42,4 +47,11 @@ def __init__(
+                    self.coordinator.data[self.entity_description.key], sensor_round
+                )
+            except TypeError as ex:
+                raise IntegrationError from ex
```
If we expect this to happen, ie it's more than a developer error, log an error instead and return None.

hesselonline replied:
OK, wil change this in new commit

cdce8p commented on homeassistant/components/wallbox/const.py:
```diff
@@ -43,6 +44,23 @@ class WallboxSensorEntityDescription(SensorEntityDescription):
+}
+
+
 SENSOR_TYPES: dict[str, WallboxSensorEntityDescription] = {
```
Could you move `SENSOR_TYPES` to `sensor.py` to address the comment from the last PR: https://github.com/home-assistant/core/pull/58690#discussion_r739772685 ?
Just need to be careful as this will likely create a merge conflict with #58801

hesselonline committed:
Update homeassistant/components/wallbox/number.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

hesselonline commented:
@MartinHjelmare @cdce8p I'm not totally convinced about moving the SENSOR_TYPES and NUMBER_TYPES to sensor.py/number.py. I understand that they are related to those components, but I feel that these are quite big dicts which really clutter up the files (at least for sensor.py). In the current approach these are 'neatly stored' in seperate const.py file.

MartinHjelmare commented:
We prefer too keep them in each platform to make it easier for readers to understand the platform and to keep the details contained in the platform.

hesselonline commented:
OK, if that is the agreed upon approach I will move those parts to the sensor/number files

hesselonline committed:
Commit to correct for time drift
Mainly an additional commit becasue I noticed the date in my
devcontainer was off by 2 days (which is confusing in git history)

MartinHjelmare commented on homeassistant/components/wallbox/number.py:
```diff
@@ -23,20 +45,27 @@ async def async_setup_entry(hass, config, async_add_entities):
     except InvalidAuth:
         pass
```
```suggestion
        return
```
Then we can remove the `else:` and outdent below.

hesselonline replied:
I don't think that works, the idea is here that we only want to add the number component if we don't have an InvalidAuth exception. (Functionally, if somebody is not authorized to do this -> don't add it.) So, if I put the `add_entities` outside of the else block it will always run, right? (I think replacing the pass with return works however, just not the removing else part.)

MartinHjelmare replied:
It will work. If we return when there's an exception the rest of the function body will not run.

hesselonline replied:
Of course, it returns from the entire function not just the try/except. I understand now, changed it.

MartinHjelmare reviewed (commented):
Looks good! A small comment below.

hesselonline committed:
changed try except

MartinHjelmare reviewed (approved):
Thanks!

cdce8p reviewed (approved):
LGTM, too. Thanks @hesselonline 🐬
I'll merge this. Should be easy enough to rebase #58801 after that.
